### ISSUE TITLE
[Bug] Does not seem to be working on Android 13/OneUI 5.X

### ISSUE THREAD

**Issue by lividhen** [LSPosed/LSPatch#178]
# Developer edited:

See https://github.com/LSPosed/LSPatch/issues/178#issuecomment-1656085625

### Steps to reproduce/复现步骤

1. Install latest lspatch
2. Install killergram and alltrans
3. Patch Firefox and telegram
4. Enable alltrans and killergram in module scope
5. Enable apps in alltrans
6. Observe nothing changing

---
1. 安装最新的lspatch
2. 安装killergram和alltrans
3. 对Firefox和telegram应用程序进行打补丁
4. 在模块范围内启用alltrans和killergram
5. 在alltrans中启用应用程序
6. 观察是否有任何变化

### Expected behaviour/预期行为

Somebody please tell me if these modules do not work under lspatch. The expected behavior is that all text in the apps is translated and sponsored messages in telegram are blocked.

---
请问有人能告诉我这些模块在lspatch下是否无法工作？预期行为是应用程序中的所有文本都被翻译，并且Telegram中的赞助消息被阻止。

### Actual behaviour/实际行为

Nothing is affected.
---
什么也没有受到影响。

### Xposed Module List/Xposed 模块列表

```Shell
AllTrans
Killergram
```

### LSPatch version/LSPatch 版本

6712

### Android version/Android 版本

13 / OneUI 5.0&5.1

### Shizuku version/Shizuku 版本

13.2, adb

### Version requirement/版本要求

- [X] I am using latest debug CI version of LSPatch and enable verbose log/我正在使用最新 CI 调试版本且启用详细日志

### Logs/日志

<details>

```
Unable to find how to get log from manager. Could someone tell me how? 

无法找到如何从管理器获取日志的方法。有人能告诉我怎么做吗？
```
</details>

[Google Drive](https://drive.google.com/file/d/13QYZomnv-BsI1MD2KjbXy-E-E5PTWkkW/view?usp=sharing)

**Comment by Lesiuk** [comment]
Same issue with Galaxy S23 Ultra. Patched apk is not patching anything.

**Comment by callmejoo** [comment]
same, s23

**Comment by pyshivam** [comment]
First of all, thank you so much for this wonderful project, 

Same problem on Android 13 in Motorola G73 5G, Modules does not seem to be working.

**Comment by pyshivam** [comment]
ok so I found out why it was not working, so LSPatch is working wonderfully the problem is with Xposed modules, I have tried it with my own module and its working. 

Potential Solutions:
1. targetSdkVersion should be 27 or lower (it will fix the SharedPreferences issue on android versions greater than 8.1) 
2. remove isModule active function in your xposedModule if you have this (you are not hooking your self so this is will always say that your module is not active however apps added to LSPatch will load modules)
3. force stop victim app and try again


My module
https://github.com/pyshivam/imeimasker/releases/tag/2.0.1 

this module works with LSPatch

**Comment by pyshivam** [comment]
> > ok so I found out why it was not working, so LSPatch is working wonderfully the problem is with Xposed modules, I have tried it with my own module and its working.
> > Potential Solutions:
> > ```
> > 1. Use canary release build of LSPatch
> > 
> > 2. targetSdkVersion should be 27 or lower of your xposed module (it will fix the SharedPreferences issue on android versions greater than 8.1 only till android version 10 after that xposed module will not by default if have SharedPreferences)
> > 
> > 3. remove isModule active function in your xposedModule if you have this (you are not hooking your self so this is will always say that your module is not active however apps added to LSPatch will load modules)
> > 
> > 4. force stop victim app and try again
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > My module https://github.com/pyshivam/imeimasker/releases/tag/2.0.1
> > this module works with LSPatch
> 
> For this scenario we need have the source code from the modules right?

Yes you will need source code or you can ask the module developer to do this

**Comment by pyshivam** [comment]
> > ok so I found out why it was not working, so LSPatch is working wonderfully the problem is with Xposed modules, I have tried it with my own module and its working.
> > Potential Solutions:
> > ```
> > 1. Use canary release build of LSPatch
> > 
> > 2. targetSdkVersion should be 27 or lower of your xposed module (it will fix the SharedPreferences issue on android versions greater than 8.1 only till android version 10 after that xposed module will not by default if have SharedPreferences)
> > 
> > 3. remove isModule active function in your xposedModule if you have this (you are not hooking your self so this is will always say that your module is not active however apps added to LSPatch will load modules)
> > 
> > 4. force stop victim app and try again
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > My module https://github.com/pyshivam/imeimasker/releases/tag/2.0.1
> > this module works with LSPatch
> 
> How i can do this in Android Studio?
> 
> ```
> remove isModule active function in your xposedModule if you have this (you are not hooking your self so this is will always say that your module is not active however apps added to LSPatch will load modules)
> ```

mostly this function is present in MainActivity you have to find out where this function is present and then return true from that function

**Comment by akhilkedia** [comment]
Hello, I am the developer of AllTrans, the xposed module mentioned above.
I recently tested AllTrans on Android 13, and it works intermittently.

The issue should most likely be on AllTrans side, as I have not updated it yet for Android 13 compatibility.

That being said, I am experiencing multiple crashes/day in latest canary of LSPatch on Android 13 on S20, with a toast saying "Manager has died". This may also be caused by issues with my module, I will investigate and reply here if it is indeed an LSPatch issue.

**Comment by zinhuofc** [comment]
> Hello, I am the developer of AllTrans, the xposed module mentioned above. I recently tested AllTrans on Android 13, and it works intermittently.
> 
> The issue should most likely be on AllTrans side, as I have not updated it yet for Android 13 compatibility.
> 
> That being said, I am experiencing multiple crashes/day in latest canary of LSPatch on Android 13 on S20, with a toast saying "Manager has died". This may also be caused by issues with my module, I will investigate and reply here if it is indeed an LSPatch issue.

hey man I'm glad to see you're ok, I thought you died because there was no more news about you or update on the app, any news? I have so much love for your app and I can't live without it, I'm currently on 12 and I can't switch unless there's an update on your app

**Comment by akhilkedia** [comment]
I think I have found (and fixed) the issue regarding compatibility with Android 13! I believe this problem will be fairly common across a range of modules.

I'll release an updated build of my app with this change. All hooks are working and all apps are getting translated :) (Tested on stock Android 13 on a Samsung phone.)

The problem was indeed on the module's side and not LSPatch!

Below are the details for any future module developers who stumble here - 

tl;dr - `ContextWrapper.attachBaseContext()` often will be called with `null` context, use some other methods instead. I am using `Application.onCreate()` along with `ContextWrapper.attachBaseContext()`

Many API calls/functions in Android require a `Context` object. A somewhat common way to do this was by hooking

I have found some other modules that use some other ways, that may or may not be better  - 
1. Hook `android.app.Instrumentation.newActivity()`
2. Hook `android.app.Instrumentation.callApplicationOnCreate()` as done in [WeiJu2](https://github.com/ikws4/WeiJu2/blob/498d341e2e499e0913a70a32c87346b49a3a3450/app/src/main/java/io/ikws4/weiju/xposed/XposedInit.java#L43C9-L43C129)
```
XposedHelpers.findAndHookMethod(Instrumentation.class, "callApplicationOnCreate", Application.class, new XC_MethodHook()
```
3. Hook `android.app.Instrumentation.newApplication()` as done in [XprivacyLua](https://github.com/M66B/XPrivacyLua/blob/85a1e498d5a9dbb902ca3d83e987ae6eec377d7a/app/src/main/java/eu/faircode/xlua/XLua.java#L246C61-L246C76)
``` 
Class<?> at = Class.forName(tiramisu ? "android.app.Instrumentation" : "android.app.LoadedApk", false, lpparam.classLoader);
XposedBridge.hookAllMethods(at, tiramisu ? "newApplication" : "makeApplication", new XC_MethodHook() 
```
I have not checked these methods in detail to say whether they are better or work.

**Comment by zinhuofc** [comment]
happy happy happy

**Comment by yujincheng08** [comment]
@akhilkedia Thx.

**Comment by zinhuofc** [comment]
> Acho que encontrei (e corrigi) o problema de compatibilidade com Android 13! Acredito que esse problema será bastante comum em vários módulos.
> 
> Lançarei uma versão atualizada do meu aplicativo com essa mudança. Todos os ganchos estão funcionando e todos os aplicativos estão sendo traduzidos :) (Testado no Android 13 padrão em um telefone Samsung.)
> 
> O problema estava realmente do lado do módulo e não do LSPatch!
> 
> Abaixo estão os detalhes para quaisquer futuros desenvolvedores de módulos que tropeçarem aqui -
> 
> dr - `ContextWrapper.attachBaseContext()`muitas vezes será chamado com `null`contexto, use alguns outros métodos. estou usando `Application.onCreate()`juntamente com `ContextWrapper.attachBaseContext()`
> 
> Muitas chamadas/funções de API no Android exigem um `Context`objeto. Uma maneira um tanto comum de fazer isso era enganchar
> 
> Eu encontrei alguns outros módulos que usam outras formas, que podem ou não ser melhores -
> 
>     1. Gancho `android.app.Instrumentation.newActivity()`
> 
>     2. Gancho `android.app.Instrumentation.callApplicationOnCreate()`como feito em [ WeiJu2 ](https://github.com/ikws4/WeiJu2/blob/498d341e2e499e0913a70a32c87346b49a3a3450/app/src/main/java/io/ikws4/weiju/xposed/XposedInit.java#L43C9-L43C129)
> 
> 
> ```
> XposedHelpers.findAndHookMethod(Instrumentation.class, "callApplicationOnCreate", Application.class, new XC_MethodHook()
> ```
> 
>     3. Gancho `android.app.Instrumentation.newApplication()`como feito em [ XprivacyLua ](https://github.com/M66B/XPrivacyLua/blob/85a1e498d5a9dbb902ca3d83e987ae6eec377d7a/app/src/main/java/eu/faircode/xlua/XLua.java#L246C61-L246C76)
> 
> 
> ```
> Class<?> at = Class.forName(tiramisu ? "android.app.Instrumentation" : "android.app.LoadedApk", false, lpparam.classLoader);
> XposedBridge.hookAllMethods(at, tiramisu ? "newApplication" : "makeApplication", new XC_MethodHook() 
> ```
> 
> Não verifiquei esses métodos detalhadamente para dizer se são melhores ou funcionam.



> I think I have found (and fixed) the issue regarding compatibility with Android 13! I believe this problem will be fairly common across a range of modules.
> 
> I'll release an updated build of my app with this change. All hooks are working and all apps are getting translated :) (Tested on stock Android 13 on a Samsung phone.)
> 
> The problem was indeed on the module's side and not LSPatch!
> 
> Below are the details for any future module developers who stumble here -
> 
> tl;dr - `ContextWrapper.attachBaseContext()` often will be called with `null` context, use some other methods instead. I am using `Application.onCreate()` along with `ContextWrapper.attachBaseContext()`
> 
> Many API calls/functions in Android require a `Context` object. A somewhat common way to do this was by hooking
> 
> I have found some other modules that use some other ways, that may or may not be better -
> 
>     1. Hook `android.app.Instrumentation.newActivity()`
> 
>     2. Hook `android.app.Instrumentation.callApplicationOnCreate()` as done in [WeiJu2](https://github.com/ikws4/WeiJu2/blob/498d341e2e499e0913a70a32c87346b49a3a3450/app/src/main/java/io/ikws4/weiju/xposed/XposedInit.java#L43C9-L43C129)
> 
> 
> ```
> XposedHelpers.findAndHookMethod(Instrumentation.class, "callApplicationOnCreate", Application.class, new XC_MethodHook()
> ```
> 
>     3. Hook `android.app.Instrumentation.newApplication()` as done in [XprivacyLua](https://github.com/M66B/XPrivacyLua/blob/85a1e498d5a9dbb902ca3d83e987ae6eec377d7a/app/src/main/java/eu/faircode/xlua/XLua.java#L246C61-L246C76)
> 
> 
> ```
> Class<?> at = Class.forName(tiramisu ? "android.app.Instrumentation" : "android.app.LoadedApk", false, lpparam.classLoader);
> XposedBridge.hookAllMethods(at, tiramisu ? "newApplication" : "makeApplication", new XC_MethodHook() 
> ```
> 
> I have not checked these methods in detail to say whether they are better or work.

only fixed for lspatch
lsposad so far nothing :/
